### ISSUE TITLE
[C++][Compute] Pre-solve chunked indices before merging chunks in Sort kernels

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#44084]
### Describe the enhancement requested

In the chunked sort kernels (for ChunkedArray and Table), the most expensive step can be the recursive merge of sorted chunks after each individual chunk was sorted.

Currently, this merge step resolves chunked indices every time an access is made to read a value. This means chunked resolution is computed `O(n*log2(k))` times (where `n` is the input length and `k` is the number of chunks).

_However_, we could instead compute chunked indices after sorting the individual chunks. Then there would be no chunk resolution when merging, just direct accesses through `ResolvedChunk`s.

### Component(s)

C++

**Comment by pitrou** [comment]
cc @felipecrv

**Comment by pitrou** [comment]
Two potential downsides to this approach:
1) the size taken by those temporary `ResolvedChunk`s is twice the size of indices, hence a bigger CPU cache footprint
2) there has to be a final "reverse resolution" step where we convert back the sorted `ResolvedChunk`s into absolute indices... or we maintain those absolute indices along the `ResolvedChunk`, which implies an ever bigger cache footprint

Experimenting will tell whether this can be beneficial.

**Comment by pitrou** [comment]
I made some initial experiments on this and I came to the following conclusion:
1. The performance is a mixed bag, with some non-negligible speedups on small input sizes (32k rows in the sort benchmarks) but also apparent slowdowns on larger inputs (8M rows). This is probably a combination of
   1) allocation cost, since 16 bytes per input row are allocated for a `int64_t` pair
   2) increased memory footprint and decreased cached efficiency, both because of enlarged indices and the temporary memory area
2. Therefore, further exploration should go towards
   1) compressing resolved indices to make them fit in 64 bits (e.g. 20 bits of `chunk_index`, 44 bits of `index_in_chunk`)
   2) transforming the logical indices to physical _in place_ before merging the chunks, and transforming them back to physical in place after merging

I might dedicate some time to this.
